python/ray/serve/_private/router.py

```diff
@@ -205,45 +240,87 @@ def max_num_scheduling_tasks(self) -> int:
     def target_num_scheduling_tasks(self) -> int:
-        """Target number of scheduling tasks to be running based on pending assignments.
+        """Target number of scheduling tasks to be running based on pending requests.
 
         This will never exceed `self.max_num_scheduling_tasks`.
         """
-        return min(self.num_pending_assignments, self.max_num_scheduling_tasks)
+        return min(self.num_pending_requests, self.max_num_scheduling_tasks)
+
+    @property
+    def curr_replicas(self) -> Dict[str, ReplicaWrapper]:
+        return self._replicas
 
     def update_replicas(self, replicas: List[ReplicaWrapper]):
         """Update the set of available replicas to be considered for scheduling.
 
         When the set of replicas changes, we may spawn additional scheduling tasks
-        if there are pending assignments.
+        if there are pending requests.
         """
-        self._replicas = {r.replica_id: r for r in replicas}
-        new_replica_id_set = set(self._replicas.keys())
+        new_replicas = {}
+        new_replica_id_set = set()
+        new_multiplexed_model_id_to_replica_ids = defaultdict(set)
+        for r in replicas:
+            new_replicas[r.replica_id] = r
+            new_replica_id_set.add(r.replica_id)
+            for model_id in r.multiplexed_model_ids:
+                new_multiplexed_model_id_to_replica_ids[model_id].add(r.replica_id)
+
+        self._replicas = new_replicas
+        self._replica_id_set = new_replica_id_set
+        self._multiplexed_model_id_to_replica_ids = (
+            new_multiplexed_model_id_to_replica_ids
+        )
+        self._replicas_updated_event.set()
+        self.maybe_start_scheduling_tasks()
+
         if self._replica_id_set != new_replica_id_set:
-            self._replica_id_set = new_replica_id_set
             logger.info(
                 "Got updated replicas for deployment "
                 f"{self._deployment_name}: {new_replica_id_set}.",
                 extra={"log_to_stderr": False},
             )
 
-        self._replicas_updated_event.set()
-        self.maybe_start_scheduling_tasks()
-
     def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
         """Shim for compatibility with the existing round robin scheduler."""
         return self.update_replicas([ActorReplicaWrapper(r) for r in running_replicas])
 
+    def _get_candidate_replica_ids(
+        self,
+        blacklist_replica_ids: Set[str],
+        request_metadata: Optional[RequestMetadata] = None,
+    ) -> Set[str]:
+        """Get candidates from the current replica set excluding the blacklist.
+
+        If a model ID is present in request_metadata, any replicas that have it are
+        prioritized.
+        """
+        if (
+            request_metadata is not None
+            and request_metadata.multiplexed_model_id
+            and request_metadata.multiplexed_model_id
+            in self._multiplexed_model_id_to_replica_ids
```

Comment:
Could you directly check if `request_metadata` is in `self._multiplexed_model_id_to_replica_ids` without checking if it's non-empty first?

```suggestion
            request_metadata is not None
            and request_metadata.multiplexed_model_id
            in self._multiplexed_model_id_to_replica_ids
```